Vote in timeout msg

dmitri-perelman opened the pull request:
[Consensus] Attach votes to timeout messages.
  
Summary:
In this diff we're extending timeout message to include an optional Vote field:
* the votes are cached in `EventProcessor` to include the last vote msg of a validator
* upon timeout the cached vote message is attached to the timeout message
* upon receiving timeout message, the votes are added to the block store and new QCs can be generated

This PR appears as two commits, but the first one belongs to another PR #367 (they're touching similar pieces of code, so it's now rebased on top of it). Please review the second commit only.

Test: added a unit test that verifies vote aggregation via timeout messages
Task: ref #371

dmitri-perelman commented:
Rebase

andll commented on consensus/src/chained_bft/consensus_types/timeout_msg.rs:
```diff
@@ -49,6 +50,17 @@ impl CryptoHash for PacemakerTimeoutSerializer {
+#[derive(Debug, Fail, PartialEq)]
+pub enum TimeoutVerificationError {
+    /// The vote message carried by the timeout doesn't pass verification.
+    #[fail(display = "VoteMsgError")]
```
Display for this one should contain info from inner element too? E.g. `"{}", _0)` part

andll commented on consensus/src/chained_bft/consensus_types/timeout_msg.rs:
```diff
@@ -126,10 +158,16 @@ impl FromProto for PacemakerTimeout {
         let round = object.get_round();
         let author = Author::try_from(object.take_author())?;
         let signature = Ed25519Signature::try_from(object.get_signature())?;
+        let vote = if let Some(vote_msg) = object.vote.into_option() {
```
nit: possible alternative:
`let vote = object.vote.into_option().map(VoteMsg::from_proto)?`

dmitri-perelman replied:
I didn't find a way to apply the '?' operator on the Result that is within the Option, hence had to open it up...

dmitri-perelman commented:
* Fixed the display function of `TimeoutVerificationError`
* Added a counter to count the number of cases, in which the vote carried by a timeout message helps to form a QC.

andll commented on consensus/src/chained_bft/event_processor.rs:
```diff
@@ -656,48 +680,44 @@ impl<T: Payload, P: ProposerInfo> EventProcessor<T, P> {
-                        error!("Error inserting qc {}: {:?}", qc, e);
-                        return;
-                    }
+            counters::TIMEOUT_VOTES_FORM_QC_COUNT.inc();
```
How do we know this is timeout vote?

dmitri-perelman replied:
Ugh, right, I moved it to a separate function that is now used in other places as well. Good catch, thanks! Let me fix it.

dmitri-perelman commented:
Fixed a bug in the counter, thanks @andll !
